qiskit_ode/signals/signals.py

```diff
@@ -27,167 +28,196 @@
-            dt, samples, start_time=start_time, carrier_freq=self.carrier_freq, phase=self.phase
-        )
+    @phase.setter
+    def phase(self, phase: float):
+        """Phase setter."""
+        if type(phase) == list:
+            phase = [Array(entry).data for entry in phase]
+        self._phase = Array(phase)
+        self._phase_arg = 1j * self._phase
 
-    def __mul__(self, other):
-        return signal_multiply(self, other)
+    def envelope(self, t: Union[float, np.array, Array]) -> Union[complex, np.array, Array]:
+        """Vectorized evaluation of the envelope at time t."""
+        return self._envelope(t)
 
-    def __rmul__(self, other):
-        return signal_multiply(self, other)
+    def complex_value(self, t: Union[float, np.array, Array]) -> Union[complex, np.array, Array]:
+        """Vectorized evaluation of the complex value at time t."""
+        arg = self._carrier_arg * t + self._phase_arg
+        return self.envelope(t) * np.exp(arg)
 
-    def __add__(self, other):
-        return signal_add(self, other)
+    def __call__(self, t: Union[float, np.array, Array]) -> Array:
+        """Vectorized evaluation of the signal at time t."""
+        return np.real(self.complex_value(t))
 
-    def __radd__(self, other):
-        return signal_add(self, other)
+    def __str__(self) -> str:
+        """Return string representation."""
+        if self.name is not None:
+            return str(self.name)
 
-    def plot(self, t0: float, tf: float, n: int, axis=None):
-        """Plot the mixed signal over an interval.
+        return "Signal(carrier_freq={freq}, phase={phase})".format(
+            freq=str(self.carrier_freq), phase=str(self.phase)
+        )
 
-        Args:
-            t0: initial time
-            tf: final time
-            n: number of points to sample in interval.
-            axis (AxesSubplot): the axis to use for plotting.
-        """
-        x_vals = np.linspace(t0, tf, n)
+    def __add__(self, other: "Signal") -> "SignalSum":
+        return signal_add(self, other)
 
-        sig_vals = []
-        for x in x_vals:
-            sig_vals.append(self.value(x))
+    def __radd__(self, other: "Signal") -> "SignalSum":
+        return self.__add__(other)
 
-        if axis:
-            axis.plot(x_vals, sig_vals)
-        else:
-            plt.plot(x_vals, sig_vals)
+    def __mul__(self, other: "Signal") -> "SignalSum":
+        return signal_multiply(self, other)
 
-    def plot_envelope(self, t0: float, tf: float, n: int, axis=None):
-        """Plot the envelope over an interval.
+    def __rmul__(self, other: "Signal") -> "SignalSum":
+        return self.__mul__(other)
 
-        Args:
-            t0: initial time
-            tf: final time
-            n: number of points to sample in interval.
-            axis (AxesSubplot): the axis to use for plotting.
-        """
-        x_vals = np.linspace(t0, tf, n)
+    def __neg__(self) -> "SignalSum":
+        return -1 * self
```

Comment:
Does this have to be a `SignalSum`? Could it not be a `Signal`?

Reply:
This does actually return a `SignalSum` - the expression `-1 * self` will result in a call to `self.__rmul__`, which returns a `SignalSum`.

Reply:
Ok, sounds good.